rfcs/20210119-determinism.md

```diff
@@ -0,0 +1,75 @@
+
+### Existing Flags
+Multiple environmental variables exist today that control determinism. As part of this change, we will deprecate then remove the following:
+
+* TF_DETERMINISTIC_OPS
+* TF_CUDNN_DETERMINISTIC
+
+tf.data also has flags for determinism. The system will throw an error message if flags are out of sync i.e. if deterministic_execution_enabled is enabled but if the tf.data option is set to ‘false’, we will throw an error. (`tf.data.Options.experimental_deterministic`). We’ll also add the necessary checks for Dataset.map and Dataset.interleave.
+
+### Grappler changes
+Grappler graph optimizations may add non-deterministic behavior. In particular some optimizations will time out if they take too long to run. When determinism is enabled, these time outs will be disabled.
+
+### Random ops
+Legacy random ops, such as `tf.random.normal`, are not deterministic if no seed is set, and so such ops will raise an error when determinism is enabled. To fix, the user should set a global seed with `tf.random.set_seed`. Since most models use legacy random ops, in practice users must call `tf.random.set_seed` when enabling deterministic behavior. Alternatively, users can pass a seed to every individual random operation, but doing so is more inconvenient.
+
+As for TensorFlow 2 random number generation, `tf.random.Generator.from_non_deterministic_state` will raise an error if called when determinism is enabled. In such cases, users should check if determinism is enabled and if so, use a different generator from a deterministic source. `tf.random.get_global_generator` implicitly calls `from_non_deterministic_state` if no global generator is set, and so will also raise an error if a global generator is not set with `tf.random.set_global_generator`.
+
+Stateless random functions, such as `tf.random.stateless_normal`, are always deterministic and so will never raise determinism-related errors.
+
+In graph mode, ops will raise an error message when the random op is created. If a random op is created in graph mode without determinism being enabled but then later runs when determinism is enabled, it will also raise an error.
+
+No error will be raised if a random op or generator is run before determinism is enabled (as is true for any other op), so users should take care to enable determinism before running any random ops or generators.
+
+### Parameter Server
+Use of parameter servers adds non-deterministic behavior. In case a model constructs a ParameterServerStrategy, TensorFlow will throw an error. We’ll also document this in the documentation for the flag.
+
+### Op Review and changes
+As part of the implementation, we will review all Ops to make a determination of their behaviour (deterministic vs non-deterministic). Some of the Ops that are known to be non-deterministic include:
+
+* `tf.nn.softmax_cross_entropy_with_logits` 
+* `tf.nn.sparse_softmax_cross_entropy_with_logits` 
+* `tf.image.resize` with method=ResizeMethod.NEAREST 
```

Comment:
gradient

Reply:
for which one?

Reply:
The comment is on line 69, and referring to that op: `tf.image.resize` with `method=ResizeMethod.NEAREST`.

Note that the other two ops are loss functions, so (in training) their back-prop paths aways end up getting used. I believe, though I'm not totally certain, that the nondeterminism is introduced in the gradient of those functions too, but I didn't mention it because of this fact that they'll always be used in real models with the gradient (at least in training, even if the model is later used only in inference with no gradient).

When the exception-throwing code is added, it may end up being only added for the gradient kernel because that's the only path that is actually injecting nondeterminism. Whatever the functionality it is, the documentation will be updated to match.